Feature - Debug mode

jordond opened the pull request:
## What's new

I have stripped out the logging logic from #68.  I have created a basic simple logger using `Log.println` as discussed in #67.

etiennelenhart commented on eiffel/src/main/java/com/etiennelenhart/eiffel/viewmodel/EiffelViewModel.kt:
```diff
@@ -111,3 +153,9 @@ abstract class EiffelViewModel<S : State, A : Action>(
 }
+
+internal fun <S : State, A : Action> EiffelViewModel<S, A>.log(message: String) {
+    if (Eiffel.debugConfig.enabled && debug) {
```
If you agree with the changes from the comment in line 49, this should also be changed to `Eiffel.debugConfig.enabled && !excludeFromDebug`.

jordond commented on eiffel/src/main/java/com/etiennelenhart/eiffel/logger/ReleaseLogger.kt:
```diff
@@ -0,0 +1,11 @@
+ *
+ * An opt-in measure to ensure no debug log statements accidentally leak into production.
+ */
+object ReleaseLogger : Logger {
```
I had the debug condition messed up in my head, so this isn't actually needed.

jordond commented on eiffel/src/main/java/com/etiennelenhart/eiffel/logger/Logger.kt:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @param[logger] Lambda for logging.
+ */
+inline fun log(crossinline logger: (priority: Int, tag: String, message: String) -> Unit): Logger {
```
Yes the `crossinline` is required.

jordond commented on eiffel/src/main/java/com/etiennelenhart/eiffel/viewmodel/EiffelViewModel.kt:
```diff
@@ -36,12 +46,23 @@ abstract class EiffelViewModel<S : State, A : Action>(
     private val actionDispatcher: CoroutineDispatcher = Dispatchers.Default
 ) : ViewModel() {
 
+    open val debug: Boolean = false
```
I agree with the tweak in the logic, however I'm not totally for adding `excludeFromDebug` as a constructor parameter. I feel like it might clutter up the constructor, I'm more partial to overriding values in the body.  And if they wanted they could still override the value in the constructor.

As for it being more discoverable, I think a properly document KDoc will cover that.

That being said if you really want it as a constructor parameter, I'm not totally against it and will make the change if you want.

etiennelenhart replied:
Yeah, good point. I think I'll also move the dispatcher parameters to protected properties later, since they'll probably won't be changed often.

jordond commented:
I've (finally) addressed most of the comments, I just have one point up above.

etiennelenhart commented on eiffel/src/main/java/com/etiennelenhart/eiffel/Eiffel.kt:
```diff
@@ -0,0 +1,30 @@
+     * Enable or disable Eiffel's debug mode.
+     *
+     * You can override the [Logger.Default] by passing in your own implementation of [Logger].
+     * @sample
```
I just learned that `@sample` is used for referencing file-private sample functions. So for inline samples like this one it would be better to replace `@sample` with `Example:`

jordond replied:
Well ain't that just annoying.  From the name it sounded perfect 🥇

etiennelenhart commented on eiffel/src/main/java/com/etiennelenhart/eiffel/logger/Logger.kt:
```diff
@@ -0,0 +1,63 @@
+ * As an example you could create a [Logger] object that uses Timber or some other
+ * custom logging library.
+ *
+ * @sample
```
Same as the comment above.

etiennelenhart commented on eiffel/src/main/java/com/etiennelenhart/eiffel/logger/Logger.kt:
```diff
@@ -0,0 +1,63 @@
+/**
+ * Convenience function to build an instance of [Logger].
+ *
+ * @sample
```
Same as previous comments.

etiennelenhart commented on eiffel/src/main/java/com/etiennelenhart/eiffel/viewmodel/EiffelViewModel.kt:
```diff
@@ -36,12 +46,23 @@ abstract class EiffelViewModel<S : State, A : Action>(
     private val actionDispatcher: CoroutineDispatcher = Dispatchers.Default
 ) : ViewModel() {
 
+    open val excludeFromDebug: Boolean = false
```
Add KDoc explaining behavior of this flag.

etiennelenhart commented on eiffel/src/main/java/com/etiennelenhart/eiffel/viewmodel/EiffelViewModel.kt:
```diff
@@ -111,3 +154,9 @@ abstract class EiffelViewModel<S : State, A : Action>(
     }
 }
+
+internal fun <S : State, A : Action> EiffelViewModel<S, A>.log(message: String) {
```
I think this should either be moved next to the `Eiffel` object or simply put into `EiffelViewModel` as a `private fun` without receiver.

jordond commented:
With your recent changes to `EiffelViewModel` and `Interception`, I think it may be worth modifying what I log in the action dispatcher.

currently: 
```kotlin
channel.consumeEach { action ->
    val currentState = _state.value!!
    log(
        """
        ->
        Processing Action:
            State: $currentState
            Action: $action
    """.trimIndent()
    )

    val resultingAction = applyInterceptions(currentState, action)
    resultingAction?.let { applyUpdate(currentState, it) }
}
```

Maybe it should be something like:
```kotlin
channel.consumeEach { action ->
    val currentState = _state.value!!
    log("Dispatching Action: $action")

    val resultingAction = applyInterceptions(currentState, action)
    resultingAction?.let { applyUpdate(currentState, it) }

    if (resultingAction != null) {
        log(
            """
            ->
            Processed Action:
                Current Action: $action
                Current State: $currentState
                Updated Action: $resultingAction
                Updated State: ${state.value}
        """.trimIndent()
        )
    }
}
```

Then maybe remove the logging from the `applyUpdate`?

Thoughts?

etiennelenhart commented:
Really nice addition, thank you!

I just noticed that you're using your 100 character line limit. I'm currently at 160 since it perfectly fits my MacBook Pro display. 😅But I realized that this is way too long and will reduce this to Android Studio's default of 120 for Eiffel. You don't have to use this for PRs though, just wanted to let you know.

jordond commented:
Yeah I'm using `ktlint` and by default it enforces a 100 character line limit, apparently it's in the kotlin standards or something.  And my IDE is setup to format on save, trying to figure out how to disable it on a per-project basis.
